### PR TITLE
feat(#46): set claim from token on user model

### LINKED ISSUE #46: Set claim from token on Django user model
We should enable pycloak to set claims from the token on the Django user model.

Let's take following example:
- claim `birthday`
- User model has a `UserProfile`, which has the field `birth_day`, i.e. `user.profile.birth_day`

We need a setting `PYCLOAK_CLAIM_TO_USER_MAPPING`, which maps the name of the claim to the field of the user. This needs to support nested related lookups, separated by dot-notation. Optionally a callback function can be specified, if the value needs parsing/processing. 

Example:
```python
from wherever import parse_birthday

PYCLOAK_CLAIM_TO_USER_MAPPING = {
    "birthday": {"field": "profile.birth_day", "callback": parse_birthday
}
```

We need a setting `PYCLOAK_CLAIM_SKIP_MISSING`. If this is `True`, pycloak skips when the claim is not present in the token and logs a warning. If this is `False`, pycloak will raise a `django.core.exceptions.ImproperlyConfigured` error.

We need a setting `PYCLOAK_CLAIM_IGNORE_VALIDATION_ERRORS`. If this is `True`, pycloak logs a warning when a validation error is raised. If this is `False`, pycloak will raise the exception.

The field on the user model is set in the middleware at `7. login user`. We only need to set it, if we login the user or if it's the same user with a new token.

### PR DESCRIPTION
**Pull request by tschale**
closes #46

### CODE REVIEW

File: pycloak/middleware.py [hunk 1]
```diff
@@ -184,3 +186,56 @@ def allow_default_login(self, request) -> bool:
...
+    def store_claim_on_user(self, request, user):
+        for claim, attr in conf.CLAIM_TO_USER_MAPPING.items():
+            if not attr.get("field"):
+                raise ImproperlyConfigured(
+                    f"PYCLOAK_CLAIM_TO_USER_MAPPING: {claim} has no field specified"
+                )
+
+            # check for related fields on the user model
+            rel_fields = attr["field"].split(".")
+            field = rel_fields.pop(0)
```
Comment (Schille) [inline pycloak/middleware.py:202]: Pops the first item?

File: pycloak/middleware.py [hunk 2]
```diff
@@ -184,3 +186,56 @@ def allow_default_login(self, request) -> bool:
...
+            # check for callback method
+            if callback_fun := attr.get("callback"):
+                try:
+                    value = callback_fun(value)
+                except Exception as e:
+                    logger.exception(
+                        f"Callback {callback_fun} failed for claim {claim}",
+                        error=str(e),
+                    )
+                    raise ImproperlyConfigured(
```
Comment (Schille) [inline pycloak/middleware.py]: Isn't this a ValueError?

File: pycloak/middleware.py [hunk 3]
```diff
@@ -184,3 +186,56 @@ def allow_default_login(self, request) -> bool:
...
+                    except ValidationError as e:
+                        logger.warning(
+                            f"Validation error for claim {claim} on field {field}",
+                            error=str(e),
+                        )
+                        if not conf.PYCLOAK_CLAIM_IGNORE_VALIDATION_ERROR:
+                            raise e
+
+            # save the object (user or related object)
+            obj.save()
```
Comment (Schille) [inline pycloak/middleware.py:246]: This is potentially called multiple times. Maybe not ideal?

File: pycloak/middleware.py (hunk 1 above)
Comment (tschale) [reply pycloak/middleware.py:202]: Pops the last (https://docs.python.org/3/tutorial/datastructures.html):

> Remove the item at the given position in the list, and return it. If no index is specified, a.pop() removes and returns the last item in the list.

File: pycloak/middleware.py (hunk 2 above)
Comment (tschale) [reply pycloak/middleware.py]: I changed it

File: pycloak/middleware.py (hunk 3 above)
Comment (tschale) [reply pycloak/middleware.py:246]: I changed it.

**Comment by Schille** [comment]
Cool.

**Review by Schille — APPROVED** [review]

**Pull request merged**